anaconda_project/internal/streaming_popen.py

```diff
@@ -0,0 +1,147 @@
+# -*- coding: utf-8 -*-
+# ----------------------------------------------------------------------------
+# Copyright © 2017, Continuum Analytics, Inc. All rights reserved.
+#
+# The full license is in the file LICENSE.txt, distributed with this software.
+# ----------------------------------------------------------------------------
+from __future__ import absolute_import, print_function
+
+import io
+import subprocess
+from threading import Thread
+
+try:
+    from queue import Queue
+except ImportError:  # pragma: no cover (py2 only)
+    from Queue import Queue  # pragma: no cover (py2 only)
+
+from anaconda_project.internal import logged_subprocess
+
+
+# this function exists to be mocked in tests
+def _read_from_stream(stream, count):
+    return stream.read(count)
+
+
+def _read_and_queue_data(pipe, queue):
+    try:
+        while True:
+            # Reading one character at a time is ridiculous, but
+            # the problem is we want to immediately display each
+            # "." in conda's "Doing stuff....." sort of output. If
+            # we read >1 char, Python apparently will call the
+            # read() system call more than once trying to fill up
+            # our requested buffer, which is undesirable because
+            # it prevents showing streaming progress to the user.
+            data = _read_from_stream(pipe, 1)
```

Comment:
I think I could use `os.read` here to do one read() system call, but then I would be reading bytes instead of the chars I'm reading here from TextIOWrapper, and I'd have to reimplement my own equivalent of TextIOWrapper.